File: medium/105. Construct Binary Tree from Preorder and Inorder Traversal/105. Construct Binary Tree from Preorder and Inorder Traversal.ts
/**
 * Definition for a binary tree node.
 */


class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

function buildTree(preorder: number[], inorder: number[]): TreeNode | null {
    if (!preorder.length || !inorder.length)
        return null

    const head = preorder.shift()!
    const index = inorder.indexOf(head)

    const left = buildTree(preorder, inorder.slice(0, index))
    const right = buildTree(preorder, inorder.slice(index + 1,))

    return new TreeNode(head, left, right)
};

File: medium/105. Construct Binary Tree from Preorder and Inorder Traversal/105. Construct Binary Tree from Preorder and Inorder Traversal.py
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder or not inorder:
            return

        head = preorder.pop(0)
        position = inorder.index(head)

        left = self.buildTree(preorder, inorder[:position])
        right = self.buildTree(preorder, inorder[position + 1 :])

        return TreeNode(val=head, left=left, right=right)


solution = Solution()
print(solution.buildTree(preorder=[3, 9, 20, 15, 7], inorder=[9, 3, 15, 20, 7]))
print(solution.buildTree(preorder=[-1], inorder=[-1]))
